### PR TITLE
[KEP-1933] Provide hack/ testing target for static analysis.

### PR DESCRIPTION
**Pull request by PurelyApplied**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Provides a test target at `kubernetes/hack/verify-govet-levee.sh` for running [go-flow-levee](https://github.com/google/go-flow-levee) static analysis, as proposed in KEP-1933.  Prow integration in `test-infra` will follow.

**Which issue(s) this PR fixes**:

Introduces analyzer dependencies and execution script.

**Special notes for your reviewer**:

In an ideal situation, analysis would be wired into a `sh_test` Bazel target.
However, because of the separation of build and testing, and because of Bazel's sandboxing of tests, `sh_test` (and `go_test`) targets have access to neither build artifacts nor a buildable environment (e.g., a configured `GOCACHE`, etc).
As a result, any analysis depending on `buildssa` or other callgraph-capable package will suffer panic at runtime.

Such analysis has more in common with `go vet`, and indeed `go vet -vettool=...` is the standard entry-point for analyzers based on the `analysis` package.  Within Kubernetes, we apply vet via `make vet`, with a vestigial `hack/verify-govet.sh` providing redirection.

Given intent of this analysis as an explicit test, it seemed to me that a developer would want a wrapping script to run locally if errors are found during PR testing, rather than requiring wider knowledge of how to properly construct `make vet WHAT="-vettool=... -config=..."`.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
KEP-1933: [at kubernetes/enhancements](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1933-secret-logging-static-analysis).

### CODE REVIEW

File: test/instrumentation/levee/OWNERS
Comment (ehashman) [inline test/instrumentation/levee/OWNERS:5]: @logicalhan is a sig-instrumentation-approver, can we update this file? It could match the one below https://github.com/kubernetes/kubernetes/pull/94661/files#diff-7cb13a5003b89536f79c038d68c9e843

**Review by ehashman — COMMENTED** [review]
/ok-to-test
/assign
/wg security

**Comment by ehashman** [comment]
Let's try again....


/area security
/wg security-audit

File: go.sum
Comment (BenTheElder) [inline go.sum:235]: this should not be in the primary vendor / go.sum
this is a development-time dependency only?

we have a pattern for avoiding this with the hack/verify* targets. see e.g. staticcheck

Comment (BenTheElder) [reply go.sum:235]: /hold

File: test/instrumentation/levee/verify-govet-levee.sh
Comment (BenTheElder) [inline test/instrumentation/levee/verify-govet-levee.sh:1]: this should be under `hack/` like the other verify targets.

File: test/instrumentation/levee/main.go
Comment (BenTheElder) [inline test/instrumentation/levee/main.go:25]: does go-flow-levee not provide a binary for this upstream?

File: test/instrumentation/levee/verify-govet-levee.sh
Comment (BenTheElder) [inline test/instrumentation/levee/verify-govet-levee.sh:25]: what's the reason for using bazel here?
none of the other verification tools use it. we use it for some large build / unit test for caching purposes, but even that is not necessarily something we're investing in further / may remove.

importantly I don't think we can use bazel here while not polluting vendor / go.mod

File: test/instrumentation/levee/main.go
Comment (PurelyApplied) [reply test/instrumentation/levee/main.go:25]: It does, but I was struggling to get it to play nicely with the `go.mod`.  But with your other comments, it looks like I missed a useful pattern in that regard.  I'll investigate shortly.

File: hack/testdata/levee-config.json
Comment (PurelyApplied) [inline hack/testdata/levee-config.json:1]: This config is a cursory first-pass.  It will be iterated upon, and [KEP-1753](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/1753-logs-sanitization/README.md) will introduce useful field tags to drive identification of `Source` fields that should not be logged.

**Comment by PurelyApplied** [comment]
@BenTheElder I want to make sure I have the general outline correct.

* Analysis testing should be exposed as a point in `k/kubernetes`, as in the script in this PR.  The configuration here is the source of truth for analysis and the third_party version consumed.
* That test point should be targeted by a new job in `k/test-infra`.

Given the above, if I were to write some Kubernetes-specific tests to verify analysis and the correctness of our configuration, should those supplementary tests live in `k/kubernetes/hack` as well?

**Comment by PurelyApplied** [comment]
/assign @fejta

Hello, fejta!  I pinged Ben above regarding best practices for where configuration tests should live.  If you have any input on those fronts, I would appreciate it.

File: hack/tools/go.mod
Comment (PurelyApplied) [inline hack/tools/go.mod:10]: TODO: Target stationary version.

**Comment by PurelyApplied** [comment]
@ehashman Thank you for your review on KEP-1933 itself.  Any input here on the implementation?  As I'm still finding my feet in the community, would you recommend anyone else to ping on this PR?

File: hack/testdata/levee-config.json
Comment (sttts) [inline hack/testdata/levee-config.json:1]: what are these? Can we use a better name which makes it clear? Are those exceptions from govet?

Comment (PurelyApplied) [reply hack/testdata/levee-config.json:1]: Json `"Comment":` blocks added for clarity.

It's a configuration for the taint-propagation analysis being introduced, specifying where we keep credentials and where those creds should not go.  The source specification will be almost entirely replaced by field tags when they are introduced by KEP-1753.  And in fact, while I adding these comments, a collection of ~20 such PRs were opened to "Add datapolicy tags to ...".

**Comment by BenTheElder** [comment]
> verify: govet-levee | 5m50s

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/94661/pull-kubernetes-verify/1321882815230382080

Noting that this takes ~6 minutes but that overally verify is at 35m which is not bad.
This might be long enough to split out like `pull-kubernetes-typecheck` in the future, but shouldn't be necessary to start.

**Comment by spiffxp** [comment]
/test pull-kubernetes-bazel-test
flake failure for //pkg/registry/autoscaling/horizontalpodautoscaler/storage

/test pull-kubernetes-e2e-gce-ubuntu-containerd 
flake failure for Kubernetes e2e suite: [sig-apps] ReplicationController should test the lifecycle of a ReplicationController [Conformance]

Agree with @BenTheElder that leaving in verify job for now sounds ok.  If it becomes too painful, may merit its own presubmit similar to typecheck. Also, creating an optional/manually triggered presubmit for this may allow for faster iteration.

File: hack/testdata/levee-config.json
Comment (spiffxp) [inline hack/testdata/levee-config.json:1]: I would recommend a less generic name for this subdir, and giving it an OWNERS file with focused approvers so you don't have to keep hitting /hack approvers

**Review by spiffxp — CHANGES_REQUESTED** [review]
I'm ok with this, and like the additional hack script.  I have one specific concern to address for iterating on `levee-config.json`

It is a shame the tool in question doesn't support yaml, so we could comment on why the specific rules.  Adding comment fields for those might make sense for future maintainers, but I won't block, can be done as followup.

**Review by ehashman — COMMENTED** [review]
Looks okay at a glance to me, deferring to SIG Testing.

**Comment by PurelyApplied** [comment]
> It is a shame the tool in question doesn't support yaml, so we could comment on why the specific rules. Adding comment fields for those might make sense for future maintainers, but I won't block, can be done as followup.

Incidentally, I think I was working on the migration from the `json` marshaller to the `sigs.k8s.io/yaml` marhsaller when you were making your comments on Friday.  Tool reference updated and configuration converted to YAML for improved comments and readability.

File: hack/testdata/levee-config.json
Comment (PurelyApplied) [reply hack/testdata/levee-config.json]: I've added a `testdata` subdirectory to hold the configuration and related OWNERS file.  Is that how you meant your comment for a less generic subdir?  `testdata` itself is already home to 100+ files, so I didn't figure you meant that when you said "this subdir."  Alternatively, I could just update the `testdata` OWNERS with a filter on this config, if that seems better organized.

It's on the SIG Security agenda this afternoon to discuss taking ownership of this tool, but I've gone ahead and put `sig-security` as the tentative OWNERS for the config.

File: hack/testdata/levee/OWNERS
Comment (spiffxp) [inline hack/testdata/levee/OWNERS:6]: This group doesn't exist in `/OWNERS_ALIASES`.  I would recommend putting github usernames in here, and deciding later whether you want to go the route of adding and using a group alias.  You would need root approvers to update `/OWNERS_ALIASES`

File: hack/testdata/levee-config.json
Comment (spiffxp) [reply hack/testdata/levee-config.json]: Filters don't work well for approvers.

I meant instead of `/hack/testdata` which sounds like it should be owned by all kinds of different people, use something named after the tool you're owning.  `/hack/testdata/levee` is fine.  I would also have been OK with something like `/hack/levee`

**Review by spiffxp — CHANGES_REQUESTED** [review]
Very cool that you migrated this to YAML!

My only remaining concern is that you address the bot's comment: https://github.com/kubernetes/kubernetes/pull/94661/files#issuecomment-720693580

I suggested how to here: https://github.com/kubernetes/kubernetes/pull/94661/files#r516229248

File: hack/testdata/levee/OWNERS
Comment (PurelyApplied) [reply hack/testdata/levee/OWNERS:6]: Gotcha.

Is there anyone in particular you'd like to see in the OWNERS then?  You, me, and dupree?

File: go.sum
Comment (BenTheElder) [reply go.sum]: this hold was addressed @PurelyApplied -- btw ~anyone can `/hold cancel` :sweat_smile:

**Comment by BenTheElder** [comment]
Can you squash the commit history? We usually only merge PRs with logical commits, rather than the full iterative history (to aid bisecting etc., we have a *ton* of commits..)

**Comment by PurelyApplied** [comment]
Yeah, I'll do that.  I try to avoid force-pushing too much, since it strips context from the comment chains.

File: hack/testdata/levee/OWNERS
Comment (spiffxp) [inline hack/testdata/levee/OWNERS:6]: I'll accept this for now, when we update with other members, I would like to be removed

**Review by spiffxp — APPROVED** [review]
/hold cancel
/approve
/lgtm

File: hack/testdata/levee/OWNERS
Comment (PurelyApplied) [reply hack/testdata/levee/OWNERS:6]: Thanks!

**Comment by PurelyApplied** [comment]
Big money no whammy stop!

/retest

**Comment by PurelyApplied** [comment]
Once more with feeling    
/retest

**Comment by PurelyApplied** [comment]
> @PurelyApplied: The following tests failed, say ...to rerun all failed tests:

You lied to me, ci-bot.  You reran passing tests.

Guess I'll call them out one at a time.

/test pull-kubernetes-e2e-kind-ipv6
/test pull-kubernetes-conformance-kind-ga-only-parallel
/test pull-kubernetes-bazel-test
/test pull-kubernetes-integration

**Comment by PurelyApplied** [comment]
/test pull-kubernetes-e2e-kind-ipv6
--

**Comment by BenTheElder** [comment]
Passing tests must pass again against (HEAD of target branch + latest PR changes).
If something merges since the last pass, then prow will invalidate results before merge.

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/README.md (merge robot)

**Comment by PurelyApplied** [comment]
Thanks for confirming my suspicions, Ben!

**Pull request merged**
